src/librustc_mir/interpret/eval_context.rs

```diff
@@ -213,33 +185,32 @@ impl<'a, 'mir, 'tcx, M> Default for InfiniteLoopDetector<'a, 'mir, 'tcx, M>
     }
 }
 
-impl<'a, 'mir, 'tcx, M> InfiniteLoopDetector<'a, 'mir, 'tcx, M>
+impl<'mir, 'tcx, M> InfiniteLoopDetector<'mir, 'tcx, M>
     where M: Machine<'mir, 'tcx>,
-          'tcx: 'a + 'mir,
 {
     /// Returns `true` if the loop detector has not yet observed a snapshot.
     pub fn is_empty(&self) -> bool {
         self.hashes.is_empty()
     }
 
-    pub fn observe_and_analyze(
+    pub fn observe_and_analyze<'a>(
         &mut self,
         machine: &M,
-        stack: &Vec<Frame<'mir, 'tcx>>,
         memory: &Memory<'a, 'mir, 'tcx, M>,
+        stack: &[Frame<'mir, 'tcx>],
     ) -> EvalResult<'tcx, ()> {
-        let snapshot = (machine, stack, memory);
 
-        let mut fx = FxHasher::default();
-        snapshot.hash(&mut fx);
-        let hash = fx.finish();
+        let mut hcx = memory.tcx.get_stable_hashing_context();
+        let mut hasher = StableHasher::<u64>::new();
+        (machine, &memory.data, &memory.cur_frame, stack).hash_stable(&mut hcx, &mut hasher);
+        let hash = hasher.finish();
 
         if self.hashes.insert(hash) {
             // No collision
             return Ok(())
         }
 
```

Comment:
Could you add an `info!` logging here saying that we are taking a snapshot? Such a snapshot is anyway so expensive that the extra check whether we should log or not should not make a difference -- and this would help a lot when debugging mysterious CTFE slowdowns.

Reply:
Sure.